.github/workflows/ci.yml

```diff
@@ -6,13 +6,86 @@ env:
   GRADLE_OPTS: -Dhttp.keepAlive=false
 
 jobs:
+  generate-test-list:
+    runs-on: ubuntu-latest
+    outputs:
+      separateTestsNames: ${{ steps.set-matrix.outputs.separateTestsNames }}
+      testsNamesExclude: ${{ steps.set-matrix.outputs.testsNamesExclude }}
+    steps:
+    - name: Set up JDK for build and test
+      uses: actions/setup-java@v2
+      with:
+        distribution: temurin # Temurin is a distribution of adoptium
+        java-version: 17
+
+    - name: Checkout security
+      uses: actions/checkout@v2
+
+    - name: Generate list of tasks
+      id: set-matrix
+      run: |
+        echo "separateTestsNames=$(./gradlew listTasksAsJSON -q --console=plain | tail -n 1)" >> $GITHUB_OUTPUT
+
   build:
     name: build
+    needs: generate-test-list
+    strategy:
+      fail-fast: false
+      matrix:
+        jdk: [11, 17]
+        platform: ["ubuntu-latest", "windows-latest"]
+    runs-on: ${{ matrix.platform }}
+
+    steps:
+    - name: Set up JDK for build and test
+      uses: actions/setup-java@v2
+      with:
+        distribution: temurin # Temurin is a distribution of adoptium
+        java-version: ${{ matrix.jdk }}
+
+    - name: Checkout security
+      uses: actions/checkout@v2
+
+    - name: Build and Test
+      uses: gradle/gradle-build-action@v2
+      with:
+        arguments: |
+          build citest -Dbuild.snapshot=false
+          -x integrationTest
+          -x spotlessCheck
+          -x checkstyleMain
+          -x checkstyleTest
+          -x spotbugsMain
+          -x spotbugsIntegrationTest
+          -x checkstyleIntegrationTest
+          -x test
```

Comment:
_Pulling this out from another comment thread_

This command doesn't look right, I'd expect to only pass what actions should be run, not a giant list of things that shouldn't be run, this is error prone IMO.  Lets work backwards from these use case if we can

- Trigger a subset of the tests as if I was a github action runner:
      `./gradlew dlicDlsflsTest` - Pretty sure this is what you are doing now, but when I look at the way the gradle task is constructed there is alot of -x'ed items

- Trigger all tests subsets in parallel (on local machine):
       `./gradlew test`
This would use the gradle / junit task parallelization to run

- **[Nice to Have]** Trigger single test:
     `./gradlew test --tests org.opensearch.security.my.test`, this would only trigger that test alone, no matter what other tasks are in play. This is broken today because the openssl tasks run even with the filter condition. I understand if this is out of scope and I'll write up a separate issue, but I'd also like to be sure this problem isn't made worse with this change,

Reply:
> Trigger a subset of the tests as if I was a github action runner:

This is possible.

> Trigger all tests subsets in parallel (on local machine):

This is also now possible. We have achieved this by splitting off the `citest` task to keep `test` for exclusive use on local machines.

>  Trigger single test:

This is also possible. Since the restoration of the openssltest the old issue applies again that this test is also executed in addition.

But coming to the `-x` exclusions in general: These are necessary because the CI task uses `gradle build test`. `build` pulls in a lot of additional task dependencies. However, I personally see spontaneously no reason why `build` is necessary, we are just using this because it was used this way before. But I think we should try to change

```
build citest -Dbuild.snapshot=false
          -x integrationTest
          -x spotlessCheck
          -x checkstyleMain
          -x checkstyleTest
          -x spotbugsMain
          -x spotbugsIntegrationTest
          -x checkstyleIntegrationTest
          -x test
```

into 

```
citest -Dbuild.snapshot=false
```

and see if this works.

Reply:
@nibix It will work:
```
root@feaf17923c58:/app# ./gradlew citest --dry-run
=======================================
OpenSearch Build Hamster says Hello!
  Gradle Version        : 8.1.1
  OS Info               : Linux 6.3.8-200.fc38.x86_64 (amd64)
  JDK Version           : 17 (Eclipse Temurin JDK)
  JAVA_HOME             : /root/.sdkman/candidates/java/17.0.7-tem
  Random Testing Seed   : 3B71B6CF19C67EF0
  In FIPS 140 mode      : false
=======================================
:compileJava SKIPPED
:processResources SKIPPED
:classes SKIPPED
:compileTestJava SKIPPED
:copyPluginPropertiesTemplate SKIPPED
:pluginProperties SKIPPED
:processTestResources SKIPPED
:testClasses SKIPPED
:copyExtraTestResources SKIPPED
:citest SKIPPED
:opensslTest SKIPPED
:test SKIPPED
:jacocoTestReport SKIPPED

Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

See https://docs.gradle.org/8.1.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD SUCCESSFUL in 786ms
```
Unfortunately we have dependency between `jacocoTestReport` and `test` and that's why we had to use `-x test` in the first place.

Reply:
Any progress on this thread?

Reply:
@peternied The `jacocoTestReport` seems to force explicit dependencies on all test tasks. See the execution log above, the `citest` triggers `jacocoTestReport` which in turn triggers `test` and `opensslTest` as a dependency. That makes the exclusions necessary. So far, we did not find a reasonable way to avoid that.

Reply:
Could an issue be made to track this for circling back onto at a later time?

Reply:
Done: https://github.com/opensearch-project/security/issues/2913